src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicTlsSecret.cs

```diff
@@ -0,0 +1,94 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.IO;
+using System.Runtime.InteropServices;
+using System.Text;
+using Microsoft.Quic;
+using static Microsoft.Quic.MsQuic;
+
+namespace System.Net.Quic;
+
+internal sealed class MsQuicTlsSecret : IDisposable
+{
+    private static readonly int ClientRandomOffset = (int)Marshal.OffsetOf(typeof(QUIC_TLS_SECRETS), "ClientRandom");
+    private static readonly int ClientEarlyTrafficSecretOffset = (int)Marshal.OffsetOf(typeof(QUIC_TLS_SECRETS), "ClientEarlyTrafficSecret");
+    private static readonly int ClientHandshakeTrafficSecretOffset = (int)Marshal.OffsetOf(typeof(QUIC_TLS_SECRETS), "ClientHandshakeTrafficSecret");
+    private static readonly int ServerHandshakeTrafficSecretOffset = (int)Marshal.OffsetOf(typeof(QUIC_TLS_SECRETS), "ServerHandshakeTrafficSecret");
+    private static readonly int ClientTrafficSecretOffset = (int)Marshal.OffsetOf(typeof(QUIC_TLS_SECRETS), "ClientTrafficSecret0");
+    private static readonly int ServerTrafficSecretOffset = (int)Marshal.OffsetOf(typeof(QUIC_TLS_SECRETS), "ServerTrafficSecret0");
+
+    public QUIC_TLS_SECRETS tlsSecret;
+    private GCHandle gcHandle;
+
+    public unsafe MsQuicTlsSecret(MsQuicContextSafeHandle handle)
+    {
+        tlsSecret = default;
+        gcHandle = GCHandle.Alloc(this, GCHandleType.Pinned);
```

Comment:
The pinning of this is to prevent GC moving the backing memory for `tlsSecret`, is that right?
The same thing could be achieved with `NativeMemory.AllocZeroed`, is that right? Just making sure I understand this.

Reply:
yes, the goal is to prevent moves as Quic can write to it any time. I was also considering to make it part of the state but decided to keep it separate. I don't have preference and I can update it if preferable.

Reply:
It's fine this was. I was just making sure I get the intent here.

Reply:
I was thinking about it more and I decided to switch to `NativeMemory`. It feel like better fit with the way how we use it and perhaps less overhead in general. I updated Dispose to have atomic `Free`.